src/lib/optimizer/strategy/chunk_pruning_rule.cpp

```diff
@@ -21,29 +22,28 @@
 
 namespace opossum {
 
-void ChunkPruningRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
-  // we only want to follow chains of predicates
-  if (node->type != LQPNodeType::Predicate) {
-    _apply_to_inputs(node);
-    return;
-  }
-
-  DebugAssert(node->input_count() == 1, "Predicate nodes should only have 1 input");
-  // try to find a chain of predicate nodes that ends in a leaf
-  std::vector<std::shared_ptr<PredicateNode>> predicate_nodes;
-
-  // Gather PredicateNodes on top of a StoredTableNode. Ignore non-filtering and ValidateNodes.
+void ChunkPruningRule::_recurse_on_inputs(
+    const std::shared_ptr<AbstractLQPNode>& node,
+    std::unordered_map<std::shared_ptr<StoredTableNode>, std::vector<PredicateChain>>& predicates_for_table_nodes,
+    PredicateChain predicate_chain) const {
```

Comment:
right. however, visit_lqp is not that easy to use here, since we don't want to visit all nodes but follow all paths. 
however, i refactored the chunk pruning rule in a way, that removes the custom recursion by first finding all StoredTableNodes (by using visit_lqp) and then gathering predicate chains on the way up. Since we are now approaching joins from the output side, we never have to branch and thus need no recursion anymore.